src/main/java/com/epam/processor/DataProcessor.java

```diff
@@ -93,31 +115,46 @@ public RoadAccident getAccidentByIndex(String index){
      * @return
      */
     public Collection<RoadAccident> getAccidentsByLocation(float minLongitude, float maxLongitude, float minLatitude, float maxLatitude){
-        return null;
+        return roadAccidentList.stream()
+                .filter(accident -> isInScope(accident.getLongitude(), minLongitude, maxLongitude)
+                        && isInScope(accident.getLatitude(), minLatitude, maxLatitude))
+                .collect(Collectors.toList());
+    }
+
+    private boolean isInScope(float actualValue, float minValue, float maxValue){
+        return minValue <= actualValue && actualValue <= maxValue;
     }
 
     /**
      * find the weather conditions which caused max number of incidents
      * @return
      */
     public List<String> getTopThreeWeatherCondition(){
-        return null;
+        Map<String, Long> groupbyWeather = roadAccidentList.stream()
+                .collect(Collectors.groupingBy(RoadAccident::getWeatherConditions, Collectors.counting()));
+        return groupbyWeather.keySet().stream()
+                .sorted((w1,w2) -> groupbyWeather.get(w2).compareTo(groupbyWeather.get(w1)))
+                .limit(3)
+                .collect(Collectors.toList());
     }
 
     /**
      * count incidents by road surface conditions
      * @return
      */
     public Map<String, Long> getCountByRoadSurfaceCondition(){
-        return null;
+        return roadAccidentList.stream()
+                .collect(Collectors.groupingBy(RoadAccident::getRoadSurfaceConditions, Collectors.counting()));
     }
 
     /**
      * To match streaming operations result, return type is a java collection instead of multimap
      * @return
      */
     public Map<String, List<String>> getAccidentIdsGroupedByAuthority(){
-        return null;
+        return roadAccidentList.stream()
+                .map(roadAccident -> roadAccident.getDistrictAuthority())
+                .collect(Collectors.groupingBy(String::toString, Collectors.toList()));
```

Comment:
Yea - you passed my tests, but this is not doing what's supposed to. At collect time you don't have information about accidentId anymore, so think about what your Collectors.toList is actually collecting? :smirk: